Feature: Request Hooks

tylergannon opened the pull request:
MCPServer is upgraded with the capacity to notify before and/or after requests run.  This will be really helpful:

* Allows collecting statistics on various types of actions, and this allows to avoid having to add analytics code into every tool definition.
* This opens up the possibility of knowing how often resources or prompts are used.
* Also enables the MCP implementation to report on which agent has invoked it, and possibly behave differently in each case (Cursor vs Claude Desktop, etc).

Developers can choose Before / After events, and can submit callbacks that are notified during EVERY request lifecycle, or else specific request methods.  The specific request method callbacks receive typed message and/or result objects.

A special OnError callback is also provided.

To accomplish this with a minimal margin for error, a bit of a refactoring was done:

- The `MCPServer.HandleMessage` method was moved from server.go into its own file.
- The individual request handlers' prototypes were changed. rather than returning `mcp.JSONRPCMessage`, they return a tuple of their specific result type and error.
- A code generation schema was introduced, to facilitate adding functionality to each clause in the switch statement within `HandleMessage()`.
- Now that HandleMessage() now receives a typed result object and possible error, it is able to pass those results to any configured callbacks.

I considered several options including a strategy pattern to make HandleMessage() more DRY, but opted to stick with the minimal reorganization of the code, and to keep the algorithm as flat and obvious as possible.  I think the codegen mitigates the repetitiveness of that section of code.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced an extensible hooks system for enhanced monitoring, logging, and error reporting during request processing.
	- Improved handling of incoming messages, resulting in a more robust server response mechanism.
- **Documentation**
	- Added a new “Extras” section to the user guide outlining how to configure and leverage request hooks.
- **Tests**
	- Expanded test coverage to ensure reliable hook execution and to validate the improved error handling mechanisms.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tylergannon commented on server/callbacks.go:
```diff
@@ -0,0 +1,350 @@
+type OnBeforeCallToolFunc func(id any, message *mcp.CallToolRequest)
+type OnAfterCallToolFunc func(id any, message *mcp.CallToolRequest, result *mcp.CallToolResult)
+
+type Callbacks struct {
+	OnBeforeAny                   []OnBeforeAnyCallbackFunc
+	OnAfterAny                    []OnAfterAnyCallbackFunc
+	OnError                       []OnErrorCallbackFunc
+	OnBeforeInitialize            []OnBeforeInitializeFunc
+	OnAfterInitialize             []OnAfterInitializeFunc
+	OnBeforePing                  []OnBeforePingFunc
+	OnAfterPing                   []OnAfterPingFunc
+	OnBeforeListResources         []OnBeforeListResourcesFunc
+	OnAfterListResources          []OnAfterListResourcesFunc
+	OnBeforeListResourceTemplates []OnBeforeListResourceTemplatesFunc
+	OnAfterListResourceTemplates  []OnAfterListResourceTemplatesFunc
+	OnBeforeReadResource          []OnBeforeReadResourceFunc
+	OnAfterReadResource           []OnAfterReadResourceFunc
+	OnBeforeListPrompts           []OnBeforeListPromptsFunc
+	OnAfterListPrompts            []OnAfterListPromptsFunc
+	OnBeforeGetPrompt             []OnBeforeGetPromptFunc
+	OnAfterGetPrompt              []OnAfterGetPromptFunc
+	OnBeforeListTools             []OnBeforeListToolsFunc
+	OnAfterListTools              []OnAfterListToolsFunc
+	OnBeforeCallTool              []OnBeforeCallToolFunc
+	OnAfterCallTool               []OnAfterCallToolFunc
+}
```
This is why the callbacks are configured only during initialization and not during operation.

ezynda3 commented:
Hi @tylergannon I really like this idea and think it's a killer feature to have. With the risk of being nitpicky though, could we rename this to `Hooks` instead of `Callbacks`? Just a preference and think it matches with the naming I've seen in other projects as well.

tylergannon commented:
@ezynda3 Done 👍🏽   Also added additional documentation comments and a more structured error messaging to allow the OnError callback better responses:

# Add error type handling and documentation for MCP server

This commit enhances the MCP server's error handling capabilities by:

1. Improving error propagation through the OnError hook system
2. Adding comprehensive documentation with usage examples
3. Ensuring error types can be properly inspected with Go's standard error handling patterns

## Error Types and Usage

We've added better support for typed errors that propagate through the request handling chain:

- `ErrUnsupported`: Used when a capability is not enabled on the server
- `UnparseableMessageError`: Used when parsing a message fails
- `ErrResourceNotFound`: Used when a requested resource doesn't exist
- `ErrPromptNotFound`: Used when a requested prompt doesn't exist
- `ErrToolNotFound`: Used when a requested tool doesn't exist

These errors can be interrogated using `errors.Is` and `errors.As` to provide more targeted error handling.

## Code Examples

The documentation now includes examples like:

```go
hooks.AddOnError(func(id any, method mcp.MCPMethod, message any, err error) {
  // Check for specific error types using errors.Is
  if errors.Is(err, ErrUnsupported) {
    // Handle capability not supported errors
    log.Printf("Capability not supported: %v", err)
  }

  // Use errors.As to get specific error types
  var parseErr = &UnparseableMessageError{}
  if errors.As(err, &parseErr) {
    // Access specific methods/fields of the error type
    log.Printf("Failed to parse message for method %s: %v", 
               parseErr.GetMethod(), parseErr.Unwrap())
    // Access the raw message that failed to parse
    rawMsg := parseErr.GetMessage()
  }

  // Check for specific resource/prompt/tool errors
  switch {
  case errors.Is(err, ErrResourceNotFound):
    log.Printf("Resource not found: %v", err)
  case errors.Is(err, ErrPromptNotFound):
    log.Printf("Prompt not found: %v", err)
  case errors.Is(err, ErrToolNotFound):
    log.Printf("Tool not found: %v", err)
  }
})
```

We've also added examples for testing scenarios:

```go
// Create a channel to receive errors for testing
errChan := make(chan error, 1)

// Register hook to capture and inspect errors
hooks := &Hooks{}
hooks.AddOnError(func(id any, method mcp.MCPMethod, message any, err error) {
    // For capability-related errors
    if errors.Is(err, ErrUnsupported) {
        // Handle capability not supported
        errChan <- err
        return
    }

    // For parsing errors
    var parseErr = &UnparseableMessageError{}
    if errors.As(err, &parseErr) {
        // Handle unparseable message errors
        fmt.Printf("Failed to parse %s request: %v\n", 
                   parseErr.GetMethod(), parseErr.Unwrap())
        errChan <- parseErr
        return
    }

    // For resource/prompt/tool not found errors
    if errors.Is(err, ErrResourceNotFound) ||
       errors.Is(err, ErrPromptNotFound) ||
       errors.Is(err, ErrToolNotFound) {
        // Handle not found errors
        errChan <- err
        return
    }

    // For other errors
    errChan <- err
})
```

These improvements make the MCP server more robust and user-friendly by providing clear error patterns and detailed documentation.

tylergannon commented on server/hooks.go:
```diff
@@ -0,0 +1,442 @@
+	for _, hook := range c.OnAfterAny {
+		hook(id, method, &message, result)
+	}
+}
```
Thanks for calling this out.  This is fixed in 6c6275c3e584508687ce1369444866f037fd2e24.
